lib/licensed/sources/mix.rb

```diff
@@ -0,0 +1,177 @@
+      # Returns a Pathname.
+      def dep_path(name)
+        config.pwd.join("deps", name)
+      end
+
+      class LockfileParser
+
+        class ParseError < RuntimeError; end
+
+        # Top-level pattern extracting the name and Mix.SCM type
+        LINE_PATTERN = /
+          \A                 # At the beginning of input
+          \s*                # after any number of spaces
+          "(?<name>.*?)"     # capture the contents of a double-quoted string as the name
+          :\s*\{             # then skipping a colon, any number of spaces, and an opening brace,
+          :(?<scm>hex|git)   # capture the contents of a Elixir atom as the scm
+          ,\s*               # and, skipping a comma and any number of spaces,
+          (?<contents>.*)    # capture the rest of input as the contents.
+        /x
+
+        # Patterns to extract the version and repo information for each Mix.SCM type.
+        SCM_PATTERN = {
+          # The Hex Package Manager
+          "hex" => /
+            \A                # At the beginning of input
+            :[a-zA-Z0-9_]+    # after an Elixir atom,
+            ,\s*              # and skipping a comma and any number of spaces,
+            "(?<version>.*?)" # capture the contents of a double-quoted string as the version,
+            .*                # and later
+            "(?<repo>.*?)"    # capture the contents of a double-quoted string as the repo
+            \},?\s*\Z         # right before the final closing brace.
+          /x,
+
+          # Git
+          "git" => /
+            \A                # At the beginning of input
+            "(?<repo>.*?)"    # capture the contents of a double-quoted string as the repo
+            ,\s*              # and, skipping a comma and any number of spaces,
+            "(?<version>.*?)" # capture the contents of a second double-quoted string as the version.
+          /x
+        }
+
+        # Parses a mix.lock to extract raw package information.
+        #
+        # path - The path to the mix.lock as a Pathname or String.
+        #
+        # Returns an Array of Hash package entries, or raises a ParserError if
+        # unsuccessful.
+        def self.read(path)
+          lines = File.readlines(path)
+          new(lines).result
+        end
+
+        def initialize(lines)
+          @lines = lines
+        end
+
+        # Parses the input lines.
+        #
+        # Returns an Array of Hash package entries, or raises a ParseError if
+        # unsuccessful.
+        def result
+          # Ignore the first and last lines of the file (the beginning and
+          # ending of the enclosing map).
+          @lines[1..-2].map do |line|
+            parse_line(line)
+          end
+        end
+
+        private
+
+        # Parse a line from the mix.lock file.
+        #
+        # line - A line of input as a String.
+        #
+        # Returns a Hash package entry, or raises a ParserError if unsuccessful.
+        def parse_line(line)
```

Comment:
Raising the custom `ParseError` will halt execution of the program and should generally be avoided.  

1. When `match` is not found - this looks like a `Source` level error, in that the `mix.lock` format is not understood and the source should halt execution.  

   You can handle `Source` errors by raising `Licensed::Sources::Source::Error`, which is handled by halting the execution of the current source and logging the error.

2. When `data` is not found - this looks like a `Dependency` level error, in that there is a name and scm match, but further data couldn't be found.

   `Dependency` errors are generally handled by setting the error message as `error` in `package_entry` which it looks like you already handle when creating the `Dependency` object.

Reply:
1. This makes sense, will, do.
2. I think the reason I didn't do this is because `version` and `path` was listed as required attributes (somewhere in the docs?). I assume I can just set `version` to `nil`, however?